### PR TITLE
[ML] Add bwc serialization unit test scaffold

### PR DESCRIPTION
**Pull request by benwtrent**
Adds new `AbstractBWCSerializationTestCase` which provides easy scaffolding for BWC serialization unit tests. 

These are no replacement for true BWC tests (which execute actual old code). These tests do provide some good coverage for the current code when serializing to/from old versions.

I changed one of our object model serialization test classes to take advantage of this change.

I think this scaffolding could be made more general/available once we have exercised it to determine it is good enough for mass consumption within the testing framework.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: This bit seems like it belongs in the tests for `OutlierDetection`. However, we'd need to access that mutation from here too to make this test work. We could reuse the `mutateInstanceForVersion` method of other tests here (by extracting the logic in a static method, similarly to our `createRandom` methods). However, I don't see how we'd get rid of the `instanceof` here.

One thing this makes me wonder is if it'd be better to design this in a way where we have a separate test for each BWC change we're testing. Say we have for example a private test method that takes some params and we can call it from different tests where we isolate the change we're testing. That would help avoiding `mutateInstanceForVersion` become too complex.

I like the idea of having these tests a lot, don't take me wrong! I think it's great! Just pondering a few ideas on the details.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: Do these belong in the respective test classes? See below.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: This one too. Also, I wonder if it's worth having a `Builder` class for this just for this test. If we had a `DataFrameAnalyticsSourceTests.mutateInstanceForVersion` it could perform this logic just by using the constructor.

**Review by dimitris-athanasiou — COMMENTED** [review]
Really like this! Left a few comments to discuss. I'm not really certain on my comments, but I think they're worth the discussion.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: > One thing this makes me wonder is if it'd be better to design this in a way where we have a separate test for each BWC change we're testing.

We could potentially do this. But I also think that this type of separation could easily be done within the `mutateInstanceForVersion` by having separate function calls. 

My goal is to implicitly force the users of the scaffold to handle all BWC versions. Whether or not they want to. Having users write their own individual bwc tests for specific versions sort of defeats this.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: This `todo` shows that it is impossible to accurately test equality for `Regression` and `Classification`. 

When reading in from an older version, a random seed is generated. There is no way of determining which value that is so that you can accurately say "this is what the object would be if serialized from an old version".

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: I can set that up, for sure.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: I agree though, that we should attempt to have static methods for us to reuse else where if an object has multiple sub-objects that have BWC tests. I will refactor a bit to handle that :D

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: Added some tests to classification and regression analysis types. 

Also updated the scaffolding to better handle unique situations like the random seed generation.

**Comment by benwtrent** [comment]
run elasticsearch-ci/2
run elasticsearch-ci/packaging-sample-matrix-unix 
run elasticsearch-ci/packaging-sample-matrix-windows

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM However, I wonder if it's worth getting a second opinion on this.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java]: Question: I wonder if that's robust w.r.t. the "ghost" version, which does not really exist. E.g. we have a constant for `7.5.3` which will exist until `7.6.0` (or `7.6.1`?) is released. I do not recall the logic, but there is some magic version.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: why is it overridden? why is `DEFAULT_BWC_VERSIONS` not sufficient?
(if possible add the answer as code comment)

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java]: how do you ignore versions where the feature does not exist yet, e.g. `7.1`??
I think we never check that.

**Review by hendrikmuhs — COMMENTED** [review]
Looks good, I added some questions

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java]: never mind, I realized this does not matter.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/AbstractBWCSerializationTestCase.java]: after some more thought: this is benign

**Review by hendrikmuhs — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java
Comment (benwtrent) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java]: It is overridden so these tests would actually do something :). I will remove the override on backport (so the BWC tests run according to the 7.series rules) and then the override can be removed in 8.series.

**Pull request merged**
